HoprWrapper

nionis opened the pull request:
Fixes #1300

QYuQianchen commented on packages/ethereum/contracts/HoprWrapper.sol:
```diff
@@ -0,0 +1,83 @@
+    HoprToken public wxHOPR;
+
+    // total amount of xHOPR collected in this contract
+    uint256 total_xHOPR = 0;
```
Why do we need to store the xHOPR balance of the wrapper contract?

nionis replied:
no need - removing

QYuQianchen commented on packages/ethereum/contracts/HoprWrapper.sol:
```diff
@@ -0,0 +1,83 @@
+        // must be xHOPR or wxHOPR
+        require(msg.sender == address(xHOPR) || msg.sender == address(wxHOPR), "Invalid token");
+        // must not be triggered by self
+        if (operator == address(this)) return;
```
For personal understanding: Why do we check `operator` here? What is the scenario where operator is self?

nionis replied:
in this smart contract we don't need this actually since we immediately mint / burn instead of transferring - removing

QYuQianchen commented on packages/ethereum/contracts/HoprWrapper.sol:
```diff
@@ -0,0 +1,83 @@
+     * @param userData bytes extra information provided by the token holder (if any)
+     * @param operatorData bytes extra information provided by the operator (if any)
+     */
+    function tokensReceived(
```
`tokensReceived` is the hook used by ERC777 but not by ERC677 (xHOPR), but `onTokenTransfer`.

See the production code of xHOPR: proxy to https://blockscout.com/poa/xdai/address/0xf8D1677c8a0c961938bf2f9aDc3F3CFDA759A9d9/contracts
or its source code here: https://github.com/poanetwork/tokenbridge-contracts/blob/master/contracts/PermittableToken.sol

For the same reason, L.66, L.70 and L.75 don't apply.

We could implement an additional function as below.
```Solidity
    /**
     * @dev Hook for xHOPR `transfer` and `transferAndCall` function
     * @param from address token holder address
     * @param amount uint256 amount of tokens to transfer
     * @param data bytes extra information provided by the token holder (if any)
     */
    function onTokenTransfer(
        address from,
        uint256 amount,
        bytes data
    ) returns (bool success) {
        // must be xHOPR token
        require(msg.sender == address(xHOPR), "Invalid token");
        // mint wxHOPR
        wxHOPR.mint(from, amount, "", "");
        emit Wrapped(from, amount);
        return true;
    }
```

nionis replied:
uh right .. total brain fast

nionis replied:
fixed by adding `onTokenTransfer` as suggested

QYuQianchen commented on packages/ethereum/contracts/HoprWrapper.sol:
```diff
@@ -0,0 +1,83 @@
+        // must not be triggered by self
+        if (operator == address(this)) return;
+
+        if (msg.sender == address(xHOPR)) {
```
We need an additional check on the `to` field. Only when the `to` is this wrapper contract, should we `mint` or `burn`.

QYuQianchen reviewed (commented):
I left a two questions and some critical issues that must be addressed before merge.

I'd also suggest to add more tests against the "Permittable Token" (xHOPR implementation)

QYuQianchen commented:
@nionis Here is a [diagram](https://lucid.app/documents/view/173afdc4-515a-4152-ad35-d83a7db69226) for the `HoprWrapper` contract

nionis commented:
thanks for the review @QYuQianchen, turns out we can't use `transfer` since it won't trigger the hook to any non-bridge contract, we must use `transferAndCall`

I have pushed an updated that addresses all feedback, next we will need to implement some claiming function in case somebody accidentally sends tokens using `transfer`

QYuQianchen reviewed (approved):
LGTM
